transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,259 @@
+
+import java.util.concurrent.ConcurrentMap;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Simple {@link ChannelPool} implementation which will create new {@link Channel}s if someone tries to acquire
+ * a {@link Channel} but none is in the pool atm. No limit on the maximal concurrent {@link Channel}s is enforced.
+ *
+ * @param <C>   the {@link Channel} type to pool.
+ * @param <K>   the {@link ChannelPoolKey} that is used to store and lookup the {@link Channel}s.
+ */
+public class SimpleChannelPool<C extends Channel, K extends ChannelPoolKey> implements ChannelPool<C, K> {
+
+    private static final AttributeKey<ChannelPoolKey> KEY = AttributeKey.newInstance("channelPoolKey");
+    private final ConcurrentMap<ChannelPoolKey, ChannelPoolSegment<C>> pool = PlatformDependent.newConcurrentHashMap();
+    private final ChannelPoolHandler<C, K> handler;
+    private final ChannelHealthChecker<C, K> healthCheck;
+    private final ChannelPoolSegmentFactory<C> segmentFactory;
+    private final Bootstrap bootstrap;
+
+    /**
+     * Creates a new instance using the {@link ActiveChannelHealthChecker} and a {@link ChannelPoolSegmentFactory} that
+     * process things in LIFO order.
+     *
+     * @param bootstrap         the {@link Bootstrap} that is used for connections
+     * @param handler           the {@link ChannelPoolHandler} that will be notified for the different pool actions
+     */
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler<C, K> handler) {
+        this(bootstrap, handler,
+             ActiveChannelHealthChecker.<C, K>instance(), ChannelPoolSegmentFactories.<C>newLifoFactory());
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param bootstrap         the {@link Bootstrap} that is used for connections
+     * @param handler           the {@link ChannelPoolHandler} that will be notified for the different pool actions
+     * @param healthCheck       the {@link ChannelHealthChecker} that will be used to check if a {@link Channel} is
+     *                          still healty when obtain from the {@link ChannelPool}
+     * @param segmentFactory    the {@link ChannelPoolSegmentFactory} that will be used to create new
+     *                          {@link ChannelPoolSegment}s when needed
+     */
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler<C, K> handler,
+                             final ChannelHealthChecker<C, K> healthCheck,
+                             final ChannelPoolSegmentFactory<C> segmentFactory) {
+        this.handler = checkNotNull(handler, "handler");
+        this.healthCheck = checkNotNull(healthCheck, "healthCheck");
+        this.segmentFactory = checkNotNull(segmentFactory, "segmentFactory");
+        this.bootstrap = checkNotNull(bootstrap, "bootstrap").clone();
+        this.bootstrap.handler(new ChannelInitializer<C>() {
+            @SuppressWarnings("unchecked")
```

Comment:
We could probably move this annotation to the statement level? (`K key = (K) ...`) It's always good to reduce the scope of `SuppressWarnings` annotation, you know :-)